src/core/ext/filters/client_channel/resolver/dns/event_engine/event_engine_client_channel_resolver.cc

```diff
@@ -0,0 +1,535 @@
+                         Ref(DEBUG_LOCATION, "OnSRVResolved"),
+                         resolver_->name_to_resolve()),
+        resolver_->name_to_resolve(), resolver_->query_timeout_ms_);
+    GRPC_EVENT_ENGINE_DNS_TRACE(
+        "DNSResolver::%p Started resolving SRV. Handle::%s", resolver_.get(),
+        HandleToString(*srv_handle_).c_str());
+  }
+  if (resolver_->request_service_config_) {
+    txt_handle_ = event_engine_resolver_->LookupTXT(
+        absl::bind_front(&EventEngineDNSRequestWrapper::OnTXTResolved,
+                         Ref(DEBUG_LOCATION, "OnTXTResolved")),
+        resolver_->name_to_resolve(), resolver_->query_timeout_ms_);
+    GRPC_EVENT_ENGINE_DNS_TRACE(
+        "DNSResolver::%p Started resolving TXT. Handle::%s", resolver_.get(),
+        HandleToString(*txt_handle_).c_str());
+  }
+}
+
+EventEngineDNSRequestWrapper::~EventEngineDNSRequestWrapper() {
+  resolver_.reset(DEBUG_LOCATION, "dns-resolving");
+}
+
+void EventEngineDNSRequestWrapper::Orphan() {
+  {
+    MutexLock lock(&on_resolved_mu_);
+    orphaned_ = true;
+    // Event if cancellation fails here, OnResolvedLocked will return early, and
+    // the resolver will never see a completed request.
+    if (hostname_handle_.has_value()) {
+      event_engine_resolver_->CancelLookup(*hostname_handle_);
+    }
+    if (srv_handle_.has_value()) {
+      event_engine_resolver_->CancelLookup(*srv_handle_);
+    }
+    for (const auto& handle : balancer_hostname_handles_) {
+      event_engine_resolver_->CancelLookup(handle);
+    }
+    if (txt_handle_.has_value()) {
+      event_engine_resolver_->CancelLookup(*txt_handle_);
+    }
+  }
+  Unref(DEBUG_LOCATION, "Orphan");
+}
+
+void EventEngineDNSRequestWrapper::OnHostnameResolved(
+    absl::StatusOr<std::vector<EventEngine::ResolvedAddress>> addresses) {
+  absl::optional<grpc_core::Resolver::Result> result;
+  {
+    MutexLock lock(&on_resolved_mu_);
+    if (orphaned_) return;
+    hostname_handle_.reset();
+    addresses_ = addresses;
+    result = OnResolvedLocked();
+  }
+  if (result.has_value()) {
+    resolver_->OnRequestComplete(std::move(*result));
+  }
+}
+
+void EventEngineDNSRequestWrapper::OnSRVResolved(
+    std::string srv_host,
+    absl::StatusOr<std::vector<EventEngine::DNSResolver::SRVRecord>>
+        srv_records) {
+  absl::optional<grpc_core::Resolver::Result> result;
+  {
+    MutexLock lock(&on_resolved_mu_);
+    if (orphaned_) return;
+    srv_handle_.reset();
+    if (srv_records.ok()) {
+      balancer_addresses_ = grpc_core::ServerAddressList();
+      // Do a subsequent hostname query if SRV records were returned
+      for (const auto& srv_record : *srv_records) {
+        balancer_hostname_handles_.insert(
+            event_engine_resolver_->LookupHostname(
+                absl::bind_front(
+                    &EventEngineDNSRequestWrapper::OnBalancerHostnamesResolved,
+                    Ref(DEBUG_LOCATION, "OnBalancerHostnamesResolved"),
+                    srv_host),
```

Comment:
Good catch, thank you.

Reply:
The `tests/cpp/naming` tests are not. I chatted with @yijiem about that today, we have a blind spot here, and we should resolve it after this client channel resolver lands since he'll be integrating his lower-level resolver implementation against it. Similarly, we do not appear to test the case when resolving multiple SRV records, the first query fails, but some subsequent query gets results. That may be harder to control for in a test, but it is a blind spot.

Reply:
Okay.  Please let's also track this somehow to make sure we don't lose track of it.

Reply:
> Okay. Please let's also track this somehow to make sure we don't lose track of it.

I wrote a fuzzer for the client channel resolver that exercises this case. I think I'd prefer to polish and land it in a separate PR, but I ran it for a few minutes (10k iterations) and got no crashes. WIP https://github.com/grpc/grpc/commit/44e6b976f94ed1475ff0639595e21f62732da5c1

Reply:
I was speaking of the second case, where SRV resolution results would be an error followed by good data, which we identified as a crashing bug. That's a bit harder to control in a fixed test with a real DNS resolver, but easy to fuzz. The first case, checking the authority for balancer addresses, would probably be better to test directly, but we could add a rudimentary check to the fuzzer.

Reply:
Done